### ISSUE TITLE
Building with JRuby 9.2.0.0 gives a java.lang.VerifyError

### ISSUE THREAD

**Issue by robertpanzer** [asciidoctor/asciidoctorj#651]
Tried building with JRuby 9.2.0.0 on Java 8 and 10 gave a java.lang.VerifyError on every test:
```
java.lang.VerifyError: Bad return type
Exception Details:
  Location:
    org/jruby/gen/InterfaceImpl663457633.resolve_converter(Ljava/lang/String;)Lorg/jruby/RubyClass; @54: areturn
  Reason:
    Type 'org/jruby/runtime/builtin/IRubyObject' (current frame, stack[0]) is not assignable to 'org/jruby/RubyClass' (from method signature)
  Current Frame:
    bci: @54
    flags: { }
    locals: { 'org/jruby/gen/InterfaceImpl663457633', 'java/lang/String', 'org/jruby/runtime/builtin/IRubyObject', 'org/jruby/Ruby' }
    stack: { 'org/jruby/runtime/builtin/IRubyObject' }
  Bytecode:
    0x0000000: 2ab4 0015 4d2c b900 1d01 004e b200 1f2c
    0x0000010: 12cc 12cd b600 272d b600 2d2c 2cb9 0031
    0x0000020: 0100 12cd 04bd 0019 5903 2d2b b800 3753
    0x0000030: b200 3db6 0043 b0                      

	at java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.lang.Class.privateGetDeclaredConstructors(Class.java:2671)
	at java.lang.Class.getConstructor0(Class.java:3075)
	at java.lang.Class.getConstructor(Class.java:1825)
	at org.jruby.javasupport.Java.newInterfaceImpl(Java.java:1352)
	at org.jruby.java.proxies.JavaInterfaceTemplate.newInterfaceProxy(JavaInterfaceTemplate.java:285)
	at org.jruby.java.proxies.JavaInterfaceTemplate.access$100(JavaInterfaceTemplate.java:34)
	at org.jruby.java.proxies.JavaInterfaceTemplate$InterfaceProxyFactory.call(JavaInterfaceTemplate.java:233)
	at org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:194)
	at org.jruby.RubyClass.finvoke(RubyClass.java:560)
	at org.jruby.runtime.Helpers.invoke(Helpers.java:353)
	at org.jruby.javasupport.JavaUtil.convertProcToInterface(JavaUtil.java:278)
	at org.jruby.RubyBasicObject.defaultToJava(RubyBasicObject.java:898)
	at org.jruby.RubyBasicObject.toJava(RubyBasicObject.java:876)
	at org.jruby.javasupport.JavaEmbedUtils.rubyToJava(JavaEmbedUtils.java:247)
	at org.asciidoctor.internal.RubyUtils.rubyToJava(RubyUtils.java:19)
	at org.asciidoctor.internal.JRubyAsciidoctorModuleFactory.createAsciidoctorModule(JRubyAsciidoctorModuleFactory.java:27)
	at org.asciidoctor.internal.JRubyAsciidoctor.createJRubyAsciidoctorInstance(JRubyAsciidoctor.java:135)
	at org.asciidoctor.internal.JRubyAsciidoctor.create(JRubyAsciidoctor.java:72)
	at org.asciidoctor.Asciidoctor$Factory.create(Asciidoctor.java:695)
	at org.asciidoctor.WhenAnAsciidoctorClassIsInstantiatedInAnEnvironmentWithGemPath.should_have_gempath_in_ruby_env_when_created_with_gempath(WhenAnAsciidoctorClassIsInstantiatedInAnEnvironmentWithGemPath.java:60)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:114)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:57)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:66)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy1.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:108)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:146)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:128)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by robertpanzer** [comment]
Maybe we have to get rid of the duck-typing here as well and properly implement AsciidoctorModule.

**Comment by robertpanzer** [comment]
I am optimistic that I can fully get away from AsciidoctorModule and asciidoctorclass.rb on 1.6.0 and make all calls explicit.
But I am pessimistic about doing this for the current master branch (1.5) without pulling all changes from 1.6 into master. (And that would mean API changes and therefore it would break semver)

**Comment by mojavelinux** [comment]
We don't yet follow semver, but we have made promises that we wouldn't break the Java API in 1.5.x.

We've decided to skip 1.6.x and jump to 2.0.0 in Asciidoctor core. Therefore, what you might consider doing is moving forward with the 1.6.0 release for AsciidoctorJ based on core 1.5.7.1 (or 1.5.8). That way, you can communicate the API change, but still be ready to align with core in 2.0.0. That should free up the AsciidoctorJ project from being stuck on 1.5.x.

**Comment by robertpanzer** [comment]
That's awesome to hear!
I am currently considering doing another 1.5 release with the latest Asciidoctor core and JRuby 9.1 once we have the logging story finalized. I am still waiting for a kind of approval or more requirements there.
(I also want to create a ticket for JRuby, sth must have changed there, maybe it's also an error in JRuby)

Then maybe we can lift the 1.6.0 to beta once I have #652 finished. Right now I am trying to create more tests for registering Ruby extensions. This revealed some more open points, like that it's not possible atm to register block or macro processors without explicitly passing the block or macro name even though the ruby class already defines it.

**Comment by robertpanzer** [comment]
Will be fixed in JRuby 9.2.1.0.
Additionally I already merged #652 that circumvents this problem on the 1.6.0 branch.
